### DISCUSSION TITLE
[FEATURE REQUEST] ssh_auth.{present,manage}: allow defining config-file being "read-only" or permissions being set

### DISCUSSION THREAD

**Discussion by mirko** [saltstack/salt#67535] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
Sometimes you want to maintain SSH access keys for users in a centralized place, instead of the user's home dirs for various reasons, in my case, I do not want users to allow adding additional SSH public keys.

This is supported by openssh-server via the sshd_config option `AuthorizedKeysFile` which in my case is set to `/etc/ssh/authorized_keys/%u`.
Salt also supports custom config paths for maintaining `ssh_auth` instances, via `config`.
If I set `config` now to `/etc/ssh/authorized_keys/user42`, though, this file is created with the following permissions:
```
/etc/ssh/authorized_keys/user42
-rw------- 1 user42 user42  857 Apr 16 22:18 user42
```
which defeats the purpose of maintaining keys in a central location outside the user's homedirs, so they can't add/change respective files - because here they still can.

**Describe the solution you'd like**
While having `config`-files managed via `ssh_auth` being read-only to the user by default is probably not feasible (though, I suppose *if* `config` is being set, it's pretty likely the admin doesn't want the user changing the file), I'd suggest either having a dedicated flag for it (e.g. `config_ro` resulting in either 400 mask or filed owned by root) or a more generic custom permission mask which I could then set to e.g. `400`.

**Describe alternatives you've considered**
I don't see any alternative here.

**Comment by OrangeDog** [comment]
The user has to own the file, so can change the permissions to whatever they want. I don't see this being very useful.

An alternative is to use `file.managed` with `create=False` and `replace=False` to manage only the permissions.

**Comment by mirko** [comment]
> The user has to own the file, so can change the permissions to whatever they want.

Sorry, but where is this coming from? The user does neither have to own the file (root is fine e.g.) nor does he/she need to have permissions for alternation.
In lot's of contexts it's exactly wanted that the user should only be allowed to read the file (for authentication via SSH).

> The user has to own the file, so can change the permissions to whatever they want. I don't see this being very useful.

Not sure I'm getting it right, but preventing users from editing (their) authorized_keys - mainly preventing them from adding additional (malicious) keys - from my point of view is an immediate and obvious benefit security wise.

This is especially the case, if users use agent-forwarding and - while their private key itself can't be stolen from the host they agent-forwarded to - the agent can be used to create new sessions to arbitrary hosts, allowing the attacker to do whatever he/she wants as long as the session where the agent is forwarded is alive - which mostly is installing backdoors / planting attacker's keys / access credentials.

Allow me to drop the following section of matrix' post mortem analysis which might be able to show the benefit more clearly: https://matrix.org/blog/2019/05/08/post-mortem-and-remediations-for-apr-11-security-incident#it-should-be-made-as-hard-as-possible-to-add-malicious-ssh-keys

**Comment by waynew** [comment]
Just did a bit of experimenting on this - it looks like it's entirely fine for non-user to own the authorized keys file. It *does* have to be readable by the particular user, though.

It *may* be possible with the current configuration to work around this, though, with both the `file.managed` permissions, as well as using `user` as root, and `config` to specify the config file directly, e.g. `config: /etc/ssh/authorized_keys/{{ user.name }}` or some such.

**Comment by OrangeDog** [comment]
Apologies, I somehow didn't notice it said `/etc/ssh/authorized_keys` and was thinking of the `~/.ssh/authorized_keys` case.
